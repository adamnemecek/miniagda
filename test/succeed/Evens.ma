-- 2010-11-01
-- 2012-01-22 parameters gone from constructors

sized codata Stream ++(A : Set) : -Size -> Set
{ cons : [i : Size] -> (head : A) -> (tail : Stream A i) -> Stream A $i
}

-- suggested by Florent Balestrini
cofun evens : [A : Set] -> [i : Size] -> Stream A (i + i) -> Stream A i
{ evens A ($i) (cons .(i + i + 1) a (cons .(i + i) b as)) =
   cons i a (evens A i as)
}

cofun map2 : [A, B : Set] -> (A -> B) ->
             [i : Size] -> Stream A (2 * i) -> Stream B (2 * i)
{ map2 A B f ($ i) (cons .$(2 * i) a1 (cons .(2 * i) a2 as)) =
    cons $(2 * i) (f a1) (cons (2 * i) (f a2) (map2 A B f i as))
}
