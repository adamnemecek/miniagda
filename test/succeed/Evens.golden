--- opening "Evens.ma" ---
--- scope checking ---
--- type checking ---
type  Stream : ++(A : Set) -> - Size -> Set
term  Stream.cons : .[A : Set] -> .[i : Size] -> ^(head : A) -> ^(tail : Stream A i) -> < Stream.cons i head tail : Stream A $i >
term  head : .[A : Set] -> .[i : Size] -> (cons : Stream A $i) -> A
{ head [A] [i] (Stream.cons [.i] #head #tail) = #head
}
term  tail : .[A : Set] -> .[i : Size] -> (cons : Stream A $i) -> Stream A i
{ tail [A] [i] (Stream.cons [.i] #head #tail) = #tail
}
term  evens : .[A : Set] -> .[i : Size] -> Stream A (i + i) -> Stream A i
{ evens [A] $[i < #] (Stream.cons [.(i + i + 1)] a (Stream.cons [.(i + i)] b as)) = Stream.cons [i] a (evens [A] [i] as)
}
term  map2 : .[A : Set] -> .[B : Set] -> (A -> B) -> .[i : Size] -> Stream A (i + i) -> Stream B (i + i)
{ map2 [A] [B] f $[i < #] (Stream.cons [.$(i + i)] a1 (Stream.cons [.(i + i)] a2 as)) = Stream.cons [$(i + i)] (f a1) (Stream.cons [i + i] (f a2) (map2 [A] [B] f [i] as))
}
--- evaluating ---
--- closing "Evens.ma" ---
